File: app/src/main/java/com/elementary/tasks/notes/preview/NotePreviewActivity.java
package com.elementary.tasks.notes.preview;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.Toast;

import com.elementary.tasks.R;
import com.elementary.tasks.core.ThemedActivity;
import com.elementary.tasks.core.data.models.Note;
import com.elementary.tasks.core.data.models.Reminder;
import com.elementary.tasks.core.utils.AssetsUtil;
import com.elementary.tasks.core.utils.BackupTool;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.Dialogues;
import com.elementary.tasks.core.utils.MeasureUtils;
import com.elementary.tasks.core.utils.Module;
import com.elementary.tasks.core.utils.Notifier;
import com.elementary.tasks.core.utils.TelephonyUtil;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.core.view_models.notes.NoteViewModel;
import com.elementary.tasks.databinding.ActivityNotePreviewBinding;
import com.elementary.tasks.navigation.settings.images.GridMarginDecoration;
import com.elementary.tasks.notes.create.CreateNoteActivity;
import com.elementary.tasks.notes.create.NoteImage;
import com.elementary.tasks.notes.list.ImagesGridAdapter;
import com.elementary.tasks.notes.list.KeepLayoutManager;
import com.elementary.tasks.reminder.create_edit.CreateReminderActivity;

import java.io.File;
import java.util.List;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class NotePreviewActivity extends ThemedActivity {

    public static final String PREVIEW_IMAGES = "preview_image_key";

    @Nullable
    private Note mNote;
    @Nullable
    private Reminder mReminder;
    private String mId;

    private final ImagesGridAdapter mAdapter = new ImagesGridAdapter();
    private ActivityNotePreviewBinding binding;
    private NoteViewModel viewModel;

    private ProgressDialog mProgress;

    private Handler mUiHandler = new Handler(Looper.getMainLooper());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mId = getIntent().getStringExtra(Constants.INTENT_ID);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_note_preview);
        initActionBar();
        initImagesList();
        initScrollView();
        initReminderCard();

        initViewModel();
    }

    private void initViewModel() {
        viewModel = ViewModelProviders.of(this, new NoteViewModel.Factory(getApplication(), mId)).get(NoteViewModel.class);
        viewModel.note.observe(this, note -> {
            if (note != null) {
                showNote(note);
            }
        });
        viewModel.reminder.observe(this, reminder -> {
            if (reminder != null) {
                showReminder(reminder);
            }
        });
        viewModel.result.observe(this, commands -> {
            if (commands != null) {
                switch (commands) {
                    case DELETED:
                        closeWindow();
                        break;
                }
            }
        });
    }

    private void initReminderCard() {
        binding.reminderContainer.setVisibility(View.GONE);
        binding.editReminder.setOnClickListener(v -> editReminder());
        binding.deleteReminder.setOnClickListener(v -> showReminderDeleteDialog());
    }

    private void editReminder() {
        if (mReminder != null) {
            startActivity(new Intent(this, CreateReminderActivity.class)
                    .putExtra(Constants.INTENT_ID, mReminder.getUuId()));
        }
    }

    private void initImagesList() {
        mAdapter.setActionsListener((view, position, noteImage, actions) -> {
            switch (actions) {
                case OPEN:
                    openImagePreview(position);
                    break;
            }
        });
        binding.imagesList.setLayoutManager(new KeepLayoutManager(this, 6, mAdapter));
        binding.imagesList.addItemDecoration(new GridMarginDecoration(getResources().getDimensionPixelSize(R.dimen.grid_item_spacing)));
        binding.imagesList.setAdapter(mAdapter);
    }

    private void openImagePreview(int position) {
        Note note = new Note();
        note.setKey(PREVIEW_IMAGES);
        note.setImages(mAdapter.getData());
        viewModel.saveNote(note);
        startActivity(new Intent(this, ImagePreviewActivity.class)
                .putExtra(Constants.INTENT_ID, note.getKey())
                .putExtra(Constants.INTENT_POSITION, position));
    }

    private void initScrollView() {
        binding.scrollContent.getViewTreeObserver().addOnScrollChangedListener(new ViewTreeObserver.OnScrollChangedListener() {
            @Override
            public void onScrollChanged() {
                int scrollY = binding.scrollContent.getScrollY();
                if (!mNote.getImages().isEmpty()) {
                    binding.appBar.getBackground().setAlpha(getAlphaForActionBar(scrollY));
                } else {
                    binding.appBar.getBackground().setAlpha(255);
                }
            }

            private int getAlphaForActionBar(int scrollY) {
                int minDist = 0, maxDist = MeasureUtils.dp2px(NotePreviewActivity.this, 200);
                if (scrollY > maxDist) {
                    return 255;
                } else if (scrollY < minDist) {
                    return 0;
                } else {
                    return (int) ((255.0 / maxDist) * scrollY);
                }
            }
        });
    }

    private void initActionBar() {
        setSupportActionBar(binding.toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowTitleEnabled(false);
        }
        binding.toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp);
        binding.toolbar.setTitle("");
    }

    private void editNote() {
        startActivity(new Intent(NotePreviewActivity.this, CreateNoteActivity.class)
                .putExtra(Constants.INTENT_ID, mNote.getKey()));
    }

    private void moveToStatus() {
        if (mNote != null) {
            new Notifier(this).showNoteNotification(mNote);
        }
    }

    @Override
    public void onBackPressed() {
        closeWindow();
    }

    private void showNote(Note note) {
        this.mNote = note;
        if (note != null) {
            binding.noteText.setText(note.getSummary());
            binding.noteText.setTypeface(AssetsUtil.getTypeface(this, note.getStyle()));
            if (Module.isLollipop()) {
                getWindow().setStatusBarColor(getThemeUtil().getNoteDarkColor(note.getColor()));
            }
            binding.scrollContent.setBackgroundColor(getThemeUtil().getNoteLightColor(note.getColor()));
            showImages(note.getImages());
        }
    }

    private void showReminder(Reminder reminder) {
        mReminder = reminder;
        if (reminder != null) {
            String dateTime = TimeUtil.getDateTimeFromGmt(reminder.getEventTime(), getPrefs().is24HourFormatEnabled());
            binding.reminderTime.setText(dateTime);
            binding.reminderContainer.setVisibility(View.VISIBLE);
        }
    }

    private void showImages(List<NoteImage> images) {
        if (!images.isEmpty()) {
            mAdapter.setImages(images);
            binding.appBar.setBackgroundColor(getThemeUtil().getNoteColor(mNote.getColor()));
            binding.appBar.getBackground().setAlpha(0);
        } else {
            binding.appBar.setBackgroundColor(getThemeUtil().getNoteColor(mNote.getColor()));
            binding.appBar.getBackground().setAlpha(255);
        }
    }

    private void hideProgress() {
        if (mProgress != null && mProgress.isShowing()) {
            mProgress.dismiss();
        }
    }

    private void showProgress() {
        mProgress = ProgressDialog.show(this, null, getString(R.string.please_wait), true, false);
    }

    private void shareNote() {
        showProgress();
        BackupTool.CreateCallback callback = this::sendNote;
        new Thread(() -> BackupTool.getInstance().createNote(mNote, callback)).start();
    }

    private void sendNote(File file) {
        hideProgress();
        if (!file.exists() || !file.canRead()) {
            Toast.makeText(this, getString(R.string.error_sending), Toast.LENGTH_SHORT).show();
            return;
        }
        TelephonyUtil.sendNote(file, this, mNote.getSummary());
        closeWindow();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.preview_note_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                closeWindow();
                return true;
            case R.id.action_share:
                shareNote();
                return true;
            case R.id.action_delete:
                showDeleteDialog();
                return true;
            case R.id.action_status:
                moveToStatus();
                return true;
            case R.id.action_edit:
                editNote();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void closeWindow() {
        if (Module.isLollipop()) {
            mUiHandler.post(this::finishAfterTransition);
        } else {
            finish();
        }
    }

    private void showDeleteDialog() {
        AlertDialog.Builder builder = Dialogues.getDialog(this);
        builder.setMessage(getString(R.string.delete_this_note));
        builder.setPositiveButton(getString(R.string.yes), (dialog, which) -> {
            dialog.dismiss();
            if (mNote != null) viewModel.deleteNote(mNote);
        });
        builder.setNegativeButton(getString(R.string.no), (dialog, which) -> dialog.dismiss());
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void showReminderDeleteDialog() {
        AlertDialog.Builder builder = Dialogues.getDialog(this);
        builder.setMessage(R.string.delete_this_reminder);
        builder.setPositiveButton(getString(R.string.yes), (dialog, which) -> {
            dialog.dismiss();
            deleteReminder();
        });
        builder.setNegativeButton(getString(R.string.no), (dialog, which) -> dialog.dismiss());
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    private void deleteReminder() {
        if (mReminder != null) {
            viewModel.deleteReminder(mReminder);
            binding.reminderContainer.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/com/elementary/tasks/notes/preview/NotePreviewActivity.kt
package com.elementary.tasks.notes.preview

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.TypedValue
import android.view.View
import androidx.activity.enableEdgeToEdge
import androidx.core.view.get
import androidx.recyclerview.widget.LinearLayoutManager
import com.elementary.tasks.AdsProvider
import com.elementary.tasks.R
import com.elementary.tasks.core.data.Commands
import com.elementary.tasks.core.data.ui.note.UiNoteImage
import com.elementary.tasks.core.data.ui.note.UiNotePreview
import com.elementary.tasks.core.interfaces.ActionsListener
import com.elementary.tasks.core.os.PermissionFlowDelegateImpl
import com.elementary.tasks.core.utils.BuildParams
import com.elementary.tasks.core.utils.ListActions
import com.elementary.tasks.core.utils.TelephonyUtil
import com.github.naz013.ui.common.Dialogues
import com.elementary.tasks.core.utils.ui.tintOverflowButton
import com.elementary.tasks.databinding.ActivityNotePreviewBinding
import com.elementary.tasks.notes.create.CreateNoteActivity
import com.elementary.tasks.notes.preview.carousel.ImagesCarouselAdapter
import com.elementary.tasks.notes.preview.reminders.AttachedRemindersAdapter
import com.elementary.tasks.notes.preview.reminders.UiNoteAttachedReminder
import com.elementary.tasks.reminder.ReminderBuilderLauncher
import com.github.naz013.common.Permissions
import com.github.naz013.common.intent.IntentKeys
import com.github.naz013.domain.note.NoteWithImages
import com.github.naz013.feature.common.livedata.nonNullObserve
import com.github.naz013.ui.common.activity.BindingActivity
import com.github.naz013.ui.common.activity.toast
import com.github.naz013.ui.common.context.colorOf
import com.github.naz013.ui.common.context.startActivity
import com.github.naz013.ui.common.isAlmostTransparent
import com.github.naz013.ui.common.isColorDark
import com.github.naz013.ui.common.login.LoginApi
import com.github.naz013.ui.common.view.ViewUtils
import com.github.naz013.ui.common.view.applyBottomInsets
import com.github.naz013.ui.common.view.applyTopInsets
import com.github.naz013.ui.common.view.gone
import com.github.naz013.ui.common.view.visible
import com.google.android.material.carousel.CarouselLayoutManager
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.koin.core.parameter.parametersOf
import java.io.File

class NotePreviewActivity : BindingActivity<ActivityNotePreviewBinding>() {

  private var isBgDark = false

  private val adapter = ImagesCarouselAdapter()
  private val attachedRemindersAdapter = AttachedRemindersAdapter(
    onEdit = { editReminder(it.id) },
    onDetach = { viewModel.detachReminder(it.id) }
  )

  private val viewModel by viewModel<NotePreviewViewModel> { parametersOf(getId()) }
  private val reminderBuilderLauncher by inject<ReminderBuilderLauncher>()
  private val dialogues by inject<Dialogues>()

  private val uiHandler = Handler(Looper.getMainLooper())

  private val imagesSingleton by inject<ImagesSingleton>()
  private val adsProvider = AdsProvider()
  private val permissionFlowDelegate = PermissionFlowDelegateImpl(this)

  override fun inflateBinding() = ActivityNotePreviewBinding.inflate(layoutInflater)

  override fun onCreate(savedInstanceState: Bundle?) {
    enableEdgeToEdge()
    super.onCreate(savedInstanceState)

    binding.scrollView.applyBottomInsets()
    isBgDark = isDarkMode
    initActionBar()
    updateTextColors()
    initImagesList()
    initReminderCard()
    initViewModel()
    loadAds()
  }

  private fun loadAds() {
    if (!BuildParams.isPro && AdsProvider.hasAds()) {
      binding.adsCard.visible()
      adsProvider.showNativeBanner(
        binding.adsHolder,
        AdsProvider.NOTE_PREVIEW_BANNER_ID,
        R.layout.list_item_ads_hor
      ) {
        binding.adsCard.visible()
      }
    } else {
      binding.adsCard.visible()
    }
  }

  private fun getId() = intent.getStringExtra(IntentKeys.INTENT_ID) ?: ""

  private fun initViewModel() {
    lifecycle.addObserver(viewModel)
    viewModel.note.nonNullObserve(this) { showNote(it) }
    viewModel.reminders.nonNullObserve(this) { showReminders(it) }
    viewModel.result.nonNullObserve(this) { commands ->
      when (commands) {
        Commands.DELETED -> closeWindow()
        else -> {
        }
      }
    }
    viewModel.error.nonNullObserve(this) { showErrorSending() }
    viewModel.sharedFile.nonNullObserve(this) { sendNote(it.first, it.second) }
  }

  private fun initReminderCard() {
    binding.attachedRemindersList.layoutManager = LinearLayoutManager(
      this,
      LinearLayoutManager.HORIZONTAL,
      false
    )
    binding.attachedRemindersList.adapter = attachedRemindersAdapter
  }

  private fun editReminder(id: String) {
    reminderBuilderLauncher.openLogged(this) {
      putExtra(IntentKeys.INTENT_ID, id)
    }
  }

  override fun onDestroy() {
    super.onDestroy()
    adsProvider.destroy()
    adapter.actionsListener = null
  }

  private fun initImagesList() {
    adapter.actionsListener = object : ActionsListener<UiNoteImage> {
      override fun onAction(view: View, position: Int, t: UiNoteImage?, actions: ListActions) {
        when (actions) {
          ListActions.OPEN -> openImagePreview(position)
          else -> {
          }
        }
      }
    }
    binding.imagesList.layoutManager = CarouselLayoutManager()
    binding.imagesList.adapter = adapter
  }

  private fun openImagePreview(position: Int) {
    imagesSingleton.setCurrent(
      images = adapter.currentList,
      backgroundColor = viewModel.note.value?.backgroundColor ?: -1
    )
    startActivity(ImagePreviewActivity::class.java) {
      putExtra(IntentKeys.INTENT_POSITION, position)
    }
  }

  private fun initActionBar() {
    binding.appBar.applyTopInsets()
    binding.toolbar.title = ""
    binding.toolbar.setNavigationOnClickListener { closeWindow() }
    binding.toolbar.setOnMenuItemClickListener { menuItem ->
      when (menuItem.itemId) {
        R.id.action_share -> {
          shareNote()
          true
        }

        R.id.action_delete -> {
          showDeleteDialog()
          true
        }

        R.id.action_status -> {
          moveToStatus()
          true
        }

        R.id.action_edit -> {
          editNote()
          true
        }

        R.id.action_archive -> {
          viewModel.toggleArchiveFlag()
          true
        }

        else -> false
      }
    }
    updateIcons()
    updateMenu()
  }

  private fun updateIcons() {
    binding.toolbar.navigationIcon = ViewUtils.backIcon(this, isBgDark)
    binding.toolbar.tintOverflowButton(isBgDark)
  }

  private fun updateMenu(isArchived: Boolean = false) {
    val archiveActionTitle = if (isArchived) {
      getString(R.string.notes_unarchive)
    } else {
      getString(R.string.notes_move_to_archive)
    }
    binding.toolbar.menu.also { menu ->
      ViewUtils.tintMenuIcon(this, menu, 0, R.drawable.ic_fluent_edit, isBgDark)
      ViewUtils.tintMenuIcon(this, menu, 1, R.drawable.ic_fluent_heart, isBgDark)
      menu[3].setTitle(archiveActionTitle)
      menu[1].isVisible = !isArchived
      menu[2].isVisible = !isArchived
    }
  }

  private fun editNote() {
    LoginApi.openLogged(this, CreateNoteActivity::class.java) {
      putExtra(IntentKeys.INTENT_ID, viewModel.key)
    }
  }

  private fun moveToStatus() {
    val uiNotePreview = viewModel.note.value ?: return
    permissionFlowDelegate.with {
      askPermission(Permissions.POST_NOTIFICATION) {
        viewModel.showNoteInNotification(uiNotePreview.id)
      }
    }
  }

  override fun handleBackPress(): Boolean {
    closeWindow()
    return true
  }

  private fun showNote(uiNotePreview: UiNotePreview) {
    showImages(uiNotePreview.images)
    binding.noteText.text = uiNotePreview.text
    binding.noteText.typeface = uiNotePreview.typeface
    binding.noteText.setTextSize(TypedValue.COMPLEX_UNIT_SP, uiNotePreview.textSize)
    window.statusBarColor = uiNotePreview.backgroundColor
    window.navigationBarColor = uiNotePreview.backgroundColor
    binding.windowBackground.setBackgroundColor(uiNotePreview.backgroundColor)
    isBgDark = if (uiNotePreview.opacity.isAlmostTransparent()) {
      isDarkMode
    } else {
      uiNotePreview.backgroundColor.isColorDark()
    }
    updateTextColors()
    updateIcons()
    updateMenu(uiNotePreview.isArchived)
  }

  private fun updateTextColors() {
    val textColor = if (isBgDark) {
      colorOf(R.color.pureWhite)
    } else {
      colorOf(R.color.pureBlack)
    }
    binding.noteText.setTextColor(textColor)
  }

  private fun showReminders(reminders: List<UiNoteAttachedReminder>) {
    if (reminders.isNotEmpty()) {
      attachedRemindersAdapter.submitList(reminders)
      binding.attachedRemindersList.visible()
    } else {
      binding.attachedRemindersList.gone()
    }
  }

  private fun showImages(images: List<UiNoteImage>) {
    adapter.submitList(images)
  }

  private fun shareNote() {
    viewModel.shareNote()
  }

  private fun sendNote(note: NoteWithImages, file: File) {
    if (isFinishing) return
    if (!file.exists() || !file.canRead()) {
      showErrorSending()
      return
    }
    TelephonyUtil.sendNote(file, this, note.note?.summary)
  }

  private fun showErrorSending() {
    toast(R.string.error_sending)
  }

  private fun closeWindow() {
    uiHandler.post { finishAfterTransition() }
  }

  private fun showDeleteDialog() {
    val builder = dialogues.getMaterialDialog(this)
    builder.setMessage(getString(R.string.delete_this_note))
    builder.setPositiveButton(getString(R.string.yes)) { dialog, _ ->
      dialog.dismiss()
      viewModel.deleteNote()
    }
    builder.setNegativeButton(getString(R.string.no)) { dialog, _ -> dialog.dismiss() }
    builder.create().show()
  }
}
